src/System.IO.Compression/tests/ZipArchive/zip_ManualAndCompatibilityTests.cs

```diff
@@ -28,5 +28,40 @@ public class zip_ManualAndCompatabilityTests
         {
             ZipTest.IsZipSameAsDir(await StreamHelpers.CreateTempCopyStream(ZipTest.compat(withTrailing)), ZipTest.compat(withoutTrailing), ZipArchiveMode.Update, dontRequireExplicit, dontCheckTimes);
         }
+
+        /// <summary>
+        /// This test ensures that a zipfile created on one platform with a file containing potentially invalid characters elsewhere
+        /// will be interpreted based on the source OS path name validation rules. 
+        /// 
+        /// For example, the file "aa\bb\cc\dd" in a zip created on Unix should be one file "aa\bb\cc\dd" whereas the same file
+        /// in a zip created on Windows should be interpreted as the file "dd" underneath three subdirectories.
+        /// </summary>
+        [Theory]
+        [InlineData("backslashes_FromUnix.zip", "aa\\bb\\cc\\dd")]
+        [InlineData("backslashes_FromWindows.zip", "dd")]
+        [InlineData("WindowsInvalid_FromUnix.zip", "aa<b>d")]
+        public static async Task ZipWithInvalidFileNames_ParsedBasedOnSourceOS(string zipName, string fileName)
+        {
+            using (Stream stream = await StreamHelpers.CreateTempCopyStream(ZipTest.compat(zipName)))
+            using (ZipArchive archive = new ZipArchive(stream))
+            {
+                Assert.Equal(1, archive.Entries.Count);
+                ZipArchiveEntry entry = archive.Entries[0];
+                Assert.Equal(fileName, entry.Name);
+            }
+        }
+
+        [Theory]
+        [InlineData("WindowsInvalid_FromWindows.zip")]
+        [InlineData("NullCharFileName_FromWindows.zip")]
+        [InlineData("NullCharFileName_FromUnix.zip")]
+        public static async Task ZipWithInvalidFileName_ThrowsArgumentException(string zipName)
+        {
+            using (Stream stream = await StreamHelpers.CreateTempCopyStream(ZipTest.compat(zipName)))
+            using (ZipArchive archive = new ZipArchive(stream))
+            {
+                Assert.Throws<ArgumentException>(() => archive.Entries);
```

Comment:
I was primarily considering when a zip path is invalid on the current system but valid on the source system. This test is for the scenario where the zip is invalid even taking into account the platform differences (e.g. the null character is wrong regardless of where it came from).

When the current implementation sees a truly invalid path it will throw an ArgumentException to indicate that the Zip is corrupt or incorrect in some way. Do you think it would be better to return the paths from the zips and provide the "full-fidelity" or throw an error because the zip contains a path that couldn't have come from the source OS?

I can imagine an example where someone creates a zip on Windows and manually tweaks the paths to be Unix-specific expecting it to work fine but they're met with unexpected exceptions when they try to access it on Unix. It's very specific but not out of the realm of possibility.